src/timely-util/src/temporal.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utilities to efficiently store future updates.
+
+use timely::order::TotalOrder;
+use timely::progress::frontier::AntichainRef;
+use timely::progress::Timestamp;
+use timely::PartialOrder;
+
+/// Timestamp extension for totally ordered timestamps that can advance by `2^exponent`.
+pub trait BucketTimestamp: TotalOrder + Timestamp {
+    /// The number of bits in the timestamp.
+    const DOMAIN: usize = size_of::<Self>() * 8;
+    /// Advance this timestamp by `2^exponent`.
+    fn advance_by_exponent(&self, exponent: usize) -> Option<Self>;
+}
+
+pub trait Storage: Sized {
+    /// The timestamp type associated with this storage.
+    type Timestamp: BucketTimestamp;
+    /// Split self in two, based on the timestamp. The result a pair of self, where the first
+    /// element contains all data with a timestamp strictly less than `timestamp`, and the second
+    /// all other data.
+    fn split(self, timestamp: &Self::Timestamp, fuel: &mut isize) -> (Self, Self);
+}
+
+/// A sorted list of buckets, representing data bucketed by timestamp.
+///
+/// Bucket chains support three main APIs: finding buckets for a given timestamp, peeling
+/// off buckets up to a frontier, and restoring the chain property. All operations aim to be
+/// amortized logarithmic in the number of outstanding timestamps.
```

Comment:
Another data structure with logarithmic complexities in these operations (insert, find, remove) is the b-tree. What advantages does the bucket chain provide over a b-tree?

Reply:
Yup, I agree. The implementation here just came together very naturally and now needs little dependencies.

Reply:
I switched the implementation to a `BTreeMap` fwiw.